File: apps/api/routers/jobs.py
from __future__ import annotations

from uuid import UUID

from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy import func, select
from sqlalchemy.orm import Session

from apps.api.database import get_db
from apps.api.dependencies import get_current_user
from apps.api.models import Company, Job, User
from apps.api.services.ats_alignment_service import (
    ATSAlignmentServiceError,
    calculate_ats_alignment,
    get_latest_ats_alignment,
)
from apps.api.services.eligibility_service import (
    evaluate_and_persist_job_eligibility,
)
from apps.api.services.job_intelligence.service import (
    JobIntelligenceServiceError,
    generate_job_intelligence,
    get_latest_job_intelligence,
)
from apps.api.services.job_match_service import (
    JobMatchServiceError,
    calculate_job_match as calculate_job_match_service,
)
from apps.api.models import AtsAlignmentResult, JobIntelligence
from services.eligibility.contracts import EligibilityResult


router = APIRouter(
    prefix="/jobs",
    tags=["jobs"],
)


@router.get("")
def list_jobs(
    db: Session = Depends(get_db),
    search: str | None = Query(default=None),
    employment_type: str | None = Query(default=None),
    remote_type: str | None = Query(default=None),
    location: str | None = Query(default=None),
    page: int = Query(default=1, ge=1),
    page_size: int = Query(default=20, ge=1, le=100),
):
    query = (
        select(Job, Company.name)
        .outerjoin(Company, Job.company_id == Company.id)
        .where(Job.is_active.is_(True))
    )

    if search:
        search_pattern = f"%{search.strip()}%"

        query = query.where(
            Job.title.ilike(search_pattern)
            | Company.name.ilike(search_pattern)
        )

    if employment_type:
        query = query.where(
            Job.employment_type == employment_type
        )

    if remote_type:
        query = query.where(
            Job.remote_type == remote_type
        )

    if location:
        query = query.where(
            Job.location.ilike(f"%{location.strip()}%")
        )

    count_query = select(func.count()).select_from(
        query.subquery()
    )

    total = db.scalar(count_query) or 0

    offset = (page - 1) * page_size

    query = (
        query
        .order_by(
            Job.posting_date.desc().nullslast(),
            Job.id.asc(),
        )
        .offset(offset)
        .limit(page_size)
    )

    results = db.execute(query).all()

    jobs = [
        {
            "id": str(job.id),
            "title": job.title,
            "company": company_name,
            "location": job.location,
            "country": job.country,
            "remote_type": job.remote_type,
            "employment_type": job.employment_type,
            "salary_min": job.salary_min,
            "salary_max": job.salary_max,
            "salary_currency": job.salary_currency,
            "contract_duration": job.contract_duration,
            "contract_worker_type": job.contract_worker_type,
            "description": job.description,
            "requirements": job.requirements,
            "responsibilities": job.responsibilities,
            "posting_date": (
                job.posting_date.isoformat()
                if job.posting_date
                else None
            ),
            "source": job.source,
            "source_url": job.source_url,
            "application_url": job.application_url,
            "first_seen_at": job.first_seen_at.isoformat(),
            "last_seen_at": job.last_seen_at.isoformat(),
        }
        for job, company_name in results
    ]

    return {
        "jobs": jobs,
        "pagination": {
            "page": page,
            "page_size": page_size,
            "total": total,
            "total_pages": (
                (total + page_size - 1) // page_size
                if total
                else 0
            ),
        },
    }


@router.post("/{job_id}/match")
def calculate_job_match(
    job_id: str,
    resume_version_id: str | None = Query(default=None),
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """
    Calculate and persist a deterministic Job Match Score for the
    authenticated user against a specific job.

    An explicit `resume_version_id` may be supplied to calculate the
    match against that exact ResumeVersion (it must belong to the
    authenticated user). When omitted, the existing default behavior
    applies: the user's most recent Resume, preferring its master
    ResumeVersion.
    """

    try:
        job_uuid = UUID(job_id)
    except ValueError:
        raise HTTPException(
            status_code=404,
            detail="Job not found",
        )

    parsed_resume_version_id: UUID | None = None

    if resume_version_id is not None:
        try:
            parsed_resume_version_id = UUID(resume_version_id)
        except ValueError:
            raise HTTPException(
                status_code=404,
                detail="Resume version not found.",
            )

    try:
        match_record = calculate_job_match_service(
            db=db,
            current_user=current_user,
            job_id=job_uuid,
            resume_version_id=parsed_resume_version_id,
        )
    except JobMatchServiceError as exc:
        raise HTTPException(
            status_code=exc.status_code,
            detail=str(exc),
        ) from exc

    result_data = match_record.result

    return {
        "id": str(match_record.id),
        "job_id": str(match_record.job_id),
        "resume_version_id": str(match_record.resume_version_id),
        "score": match_record.score,
        "confidence": match_record.confidence,
        "engine_version": match_record.engine_version,
        "strengths": result_data["strengths"],
        "skill_gaps": result_data["skill_gaps"],
        "components": result_data["components"],
        "must_have_matches": result_data["must_have_matches"],
        "must_have_gaps": result_data["must_have_gaps"],
        "preferred_matches": result_data["preferred_matches"],
        "preferred_gaps": result_data["preferred_gaps"],
    }


def _eligibility_result_to_response(
    job_id: UUID,
    result: EligibilityResult,
    *,
    record_id: UUID,
    evaluated_at,
) -> dict:
    return {
        "id": str(record_id),
        "job_id": str(job_id),
        "status": result.status.value,
        "engine_version": result.engine_version,
        "checks": [
            {
                "constraint": check.constraint,
                "status": check.status.value,
                "reason": check.reason,
            }
            for check in result.checks
        ],
        "failed_constraints": result.failed_constraints,
        "unknown_constraints": result.unknown_constraints,
        "reasons": result.reasons,
        "evaluated_at": evaluated_at.isoformat(),
    }


@router.get("/{job_id}/eligibility")
def get_job_eligibility(
    job_id: str,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """
    Evaluate deterministic Hard Eligibility (AJI-011) for the
    authenticated user against a specific job.

    This is intentionally separate from /jobs/{job_id}/match: eligibility
    is a hard pre-filter (ELIGIBLE/INELIGIBLE/UNKNOWN with explainable
    checks), never a score. The result is recalculated query-time from
    the user's current Preference/Profile and the job's current data, and
    then upserted into JobEligibilityResult (one row per user/job,
    overwritten in place) so a later AJI-012/AJI-013 consumer can read
    the job's current hard-eligibility status without recomputing it. See
    docs/ARCHITECTURE.md for the full hard-vs-soft rationale. Public job
    browsing (GET /jobs) never exposes this personalized data; this
    endpoint always requires authentication.
    """
    try:
        job_uuid = UUID(job_id)
    except ValueError:
        raise HTTPException(
            status_code=404,
            detail="Job not found",
        )

    job = db.query(Job).filter(Job.id == job_uuid).first()

    if job is None:
        raise HTTPException(
            status_code=404,
            detail="Job not found",
        )

    result, record = evaluate_and_persist_job_eligibility(
        db=db,
        current_user=current_user,
        job=job,
    )

    return _eligibility_result_to_response(
        job.id,
        result,
        record_id=record.id,
        evaluated_at=record.updated_at,
    )


def _job_intelligence_to_response(record: JobIntelligence) -> dict:
    return {
        "id": str(record.id),
        "job_id": str(record.job_id),
        "content_fingerprint": record.content_fingerprint,
        "analysis_version": record.analysis_version,
        "analyzer_version": record.analyzer_version,
        "prompt_version": record.prompt_version,
        "model_provider": record.model_provider,
        "model_name": record.model_name,
        "extraction_status": record.extraction_status,
        "source": record.source,
        "source_url": record.source_url,
        "created_at": record.created_at.isoformat(),
        "intelligence": record.structured_intelligence,
    }


@router.get("/{job_id}/intelligence")
def get_job_intelligence(
    job_id: str,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """
    Return the most recently computed Job Intelligence (AJI-012) snapshot
    for a job, without recomputing it. 404s when no snapshot has been
    generated yet (see POST /jobs/{job_id}/intelligence).

    Job Intelligence is shared, job-scoped data — never user-specific —
    but this endpoint still requires authentication like the rest of the
    per-job API surface, and never triggers an AI call on read.
    """
    try:
        job_uuid = UUID(job_id)
    except ValueError:
        raise HTTPException(
            status_code=404,
            detail="Job not found",
        )

    job = db.query(Job).filter(Job.id == job_uuid).first()

    if job is None:
        raise HTTPException(
            status_code=404,
            detail="Job not found",
        )

    record = get_latest_job_intelligence(db, job_id=job_uuid)

    if record is None:
        raise HTTPException(
            status_code=404,
            detail="Job Intelligence has not been generated for this job yet.",
        )

    return _job_intelligence_to_response(record)


@router.post("/{job_id}/intelligence")
def create_job_intelligence(
    job_id: str,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """
    Compute (or idempotently reuse) the Job Intelligence snapshot for a
    job. Reuses an existing snapshot when the job's observable content
    and the analyzer/prompt pipeline version are unchanged; otherwise
    produces a new, additional snapshot without overwriting history.
    """
    try:
        job_uuid = UUID(job_id)
    except ValueError:
        raise HTTPException(
            status_code=404,
            detail="Job not found",
        )

    try:
        record = generate_job_intelligence(db, job_id=job_uuid)
    except JobIntelligenceServiceError as exc:
        raise HTTPException(
            status_code=exc.status_code,
            detail=str(exc),
        ) from exc

    return _job_intelligence_to_response(record)


def _ats_alignment_to_response(record: AtsAlignmentResult) -> dict:
    result_data = record.result

    return {
        "id": str(record.id),
        "job_id": str(record.job_id),
        "resume_version_id": str(record.resume_version_id),
        "job_intelligence_id": str(record.job_intelligence_id),
        "engine_version": record.engine_version,
        "overall_score": record.overall_score,
        "confidence": record.confidence,
        "scoring_version": result_data["scoring_version"],
        "must_have_total": result_data["must_have_total"],
        "must_have_matched": result_data["must_have_matched"],
        "preferred_total": result_data["preferred_total"],
        "preferred_matched": result_data["preferred_matched"],
        "requirement_results": result_data["requirement_results"],
        "created_at": record.created_at.isoformat(),
    }


def _parse_optional_resume_version_id(
    resume_version_id: str | None,
) -> UUID | None:
    if resume_version_id is None:
        return None

    try:
        return UUID(resume_version_id)
    except ValueError:
        raise HTTPException(
            status_code=404,
            detail="Resume version not found.",
        )


@router.get("/{job_id}/ats")
def get_ats_alignment(
    job_id: str,
    resume_version_id: str | None = Query(default=None),
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """
    Return the authenticated user's most recently computed ATS Alignment
    (AJI-013) result for a job, without recomputing it. 404s when no
    analysis has been generated yet (see POST /jobs/{job_id}/ats).

    ATS Alignment is user-specific (the same job can be analyzed against
    different resumes/users) — a user can only ever read their own
    results, never another user's.
    """
    try:
        job_uuid = UUID(job_id)
    except ValueError:
        raise HTTPException(
            status_code=404,
            detail="Job not found",
        )

    job = db.query(Job).filter(Job.id == job_uuid).first()

    if job is None:
        raise HTTPException(
            status_code=404,
            detail="Job not found",
        )

    parsed_resume_version_id = _parse_optional_resume_version_id(
        resume_version_id
    )

    record = get_latest_ats_alignment(
        db,
        user_id=current_user.id,
        job_id=job_uuid,
        resume_version_id=parsed_resume_version_id,
    )

    if record is None:
        raise HTTPException(
            status_code=404,
            detail="ATS Alignment has not been generated for this job yet.",
        )

    return _ats_alignment_to_response(record)


@router.post("/{job_id}/ats")
def create_ats_alignment(
    job_id: str,
    resume_version_id: str | None = Query(default=None),
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """
    Compute (or idempotently reuse) the ATS Alignment (AJI-013) result
    for the authenticated user against a specific job.

    An explicit `resume_version_id` may be supplied to align against
    that exact ResumeVersion (it must belong to the authenticated user).
    When omitted, the same default resume selection Job Match uses
    applies: the user's most recent Resume, preferring its master
    ResumeVersion.

    Reuses an existing result when the resume version, the underlying
    Job Intelligence snapshot, and the ATS engine version are all
    unchanged; otherwise produces a new, additional result without
    overwriting history.
    """
    try:
        job_uuid = UUID(job_id)
    except ValueError:
        raise HTTPException(
            status_code=404,
            detail="Job not found",
        )

    parsed_resume_version_id = _parse_optional_resume_version_id(
        resume_version_id
    )

    try:
        record = calculate_ats_alignment(
            db=db,
            current_user=current_user,
            job_id=job_uuid,
            resume_version_id=parsed_resume_version_id,
        )
    except ATSAlignmentServiceError as exc:
        raise HTTPException(
            status_code=exc.status_code,
            detail=str(exc),
        ) from exc

    return _ats_alignment_to_response(record)

File: apps/web/lib/jobs.ts
export type Job = {
  id: string;
  title: string;
  company: string | null;
  location: string | null;
  country: string;
  remote_type: string | null;
  employment_type: string | null;
  salary_min: number | null;
  salary_max: number | null;
  salary_currency: string | null;
  contract_duration: string | null;
  contract_worker_type: string | null;
  description: string | null;
  requirements: string | null;
  responsibilities: string | null;
  posting_date: string | null;
  source: string;
  source_url: string | null;
  application_url: string | null;
  first_seen_at: string;
  last_seen_at: string;
};

export type JobsResponse = {
  jobs: Job[];
  pagination: {
    page: number;
    page_size: number;
    total: number;
    total_pages: number;
  };
};

const API_BASE_URL =
  process.env.NEXT_PUBLIC_API_URL || "http://localhost:8000";

export async function getJobs(params: {
  search?: string;
  employment_type?: string;
  remote_type?: string;
  location?: string;
  page?: number;
  page_size?: number;
} = {}): Promise<JobsResponse> {
  const searchParams = new URLSearchParams();

  if (params.search) {
    searchParams.set("search", params.search);
  }

  if (params.employment_type) {
    searchParams.set("employment_type", params.employment_type);
  }

  if (params.remote_type) {
    searchParams.set("remote_type", params.remote_type);
  }

  if (params.location) {
    searchParams.set("location", params.location);
  }

  searchParams.set("page", String(params.page ?? 1));
  searchParams.set("page_size", String(params.page_size ?? 20));

  const response = await fetch(
    `${API_BASE_URL}/jobs?${searchParams.toString()}`,
    {
      cache: "no-store",
    }
  );

  if (!response.ok) {
    throw new Error(`Failed to fetch jobs: ${response.status}`);
  }

  return response.json();
}
export type SkillEvidence = {
  skill: string;
  status: "matched" | "missing" | "uncertain";
  evidence_type:
    | "explicit"
    | "experience"
    | "project"
    | "education"
    | "inferred"
    | "none";
  evidence: string | null;
};

export type MatchComponent = {
  name: string;
  score: number;
  max_score: number;
  explanation: string;
};

export type JobMatchResult = {
  id: string;
  job_id: string;
  resume_version_id: string;
  job_intelligence_id: string | null;
  score: number;
  confidence: string;
  engine_version: string;
  strengths: string[];
  skill_gaps: string[];
  components: MatchComponent[];
  must_have_matches: SkillEvidence[];
  must_have_gaps: SkillEvidence[];
  preferred_matches: SkillEvidence[];
  preferred_gaps: SkillEvidence[];
};

export type JobIntelligenceSkill = {
  canonical_skill: string;
  level: "required" | "preferred";
  evidence_text: string;
  confidence: "high" | "medium" | "low";
};

export type JobIntelligenceExperience = {
  level: "required" | "preferred";
  minimum_years: number | null;
  maximum_years: number | null;
  area: string | null;
  context: string | null;
  evidence_text: string;
  confidence: "high" | "medium" | "low";
};

export type JobIntelligenceEducation = {
  level: "required" | "preferred";
  degree_level: string | null;
  field_of_study: string | null;
  evidence_text: string;
  confidence: "high" | "medium" | "low";
};

export type JobIntelligenceCertification = {
  level: "required" | "preferred";
  name: string;
  evidence_text: string;
  confidence: "high" | "medium" | "low";
};

export type JobIntelligenceResponsibility = {
  description: string;
  evidence_text: string;
};

export type JobIntelligenceData = {
  identity: {
    original_title: string;
    normalized_title: string | null;
    role_family: string | null;
    seniority: string | null;
  };
  employment: {
    employment_type: string;
    evidence_text: string | null;
  };
  location: {
    raw_location: string | null;
    city: string | null;
    state: string | null;
    country: string | null;
    additional_locations: string[];
    remote_type: string;
    work_arrangement_text: string | null;
    relocation_mentioned: boolean;
  };
  required_skills: JobIntelligenceSkill[];
  preferred_skills: JobIntelligenceSkill[];
  required_experience: JobIntelligenceExperience[];
  preferred_experience: JobIntelligenceExperience[];
  education: JobIntelligenceEducation[];
  certifications: JobIntelligenceCertification[];
  responsibilities: JobIntelligenceResponsibility[];
  authorization: {
    sponsorship: string;
    citizenship: string;
    clearance: string;
    work_authorization: string;
  };
  compensation: {
    salary_min: number | null;
    salary_max: number | null;
    currency: string | null;
    period: string;
    evidence_text: string | null;
  };
  domain: {
    value: string | null;
    confidence: string | null;
  };
};

export type JobIntelligenceResponse = {
  id: string;
  job_id: string;
  analysis_version: string;
  extraction_status: string;
  created_at: string;
  intelligence: JobIntelligenceData;
};

export type EligibilityStatus = "eligible" | "ineligible" | "unknown";

export type EligibilityCheckStatus = "pass" | "fail" | "unknown" | "not_applicable";

export type EligibilityCheck = {
  constraint: string;
  status: EligibilityCheckStatus;
  reason: string;
};

export type JobEligibilityResult = {
  id: string;
  job_id: string;
  status: EligibilityStatus;
  engine_version: string;
  checks: EligibilityCheck[];
  failed_constraints: string[];
  unknown_constraints: string[];
  reasons: string[];
  evaluated_at: string;
};

function authHeaders(): Record<string, string> {
  const token =
    typeof window !== "undefined"
      ? localStorage.getItem("ai_job_intelligence_token")
      : null;

  return token ? { Authorization: `Bearer ${token}` } : {};
}

export async function getJobIntelligence(
  jobId: string,
): Promise<JobIntelligenceResponse> {
  const response = await fetch(`${API_BASE_URL}/jobs/${jobId}/intelligence`, {
    headers: authHeaders(),
    cache: "no-store",
  });

  const data = await response.json().catch(() => null);

  if (!response.ok) {
    const message =
      typeof data?.detail === "string"
        ? data.detail
        : "Unable to load Job Intelligence.";

    throw new Error(message);
  }

  return data as JobIntelligenceResponse;
}

export async function generateJobIntelligence(
  jobId: string,
): Promise<JobIntelligenceResponse> {
  const response = await fetch(`${API_BASE_URL}/jobs/${jobId}/intelligence`, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
      ...authHeaders(),
    },
    cache: "no-store",
  });

  const data = await response.json().catch(() => null);

  if (!response.ok) {
    const message =
      typeof data?.detail === "string"
        ? data.detail
        : "Unable to generate Job Intelligence.";

    throw new Error(message);
  }

  return data as JobIntelligenceResponse;
}

export async function getJobEligibility(
  jobId: string,
): Promise<JobEligibilityResult> {
  const response = await fetch(`${API_BASE_URL}/jobs/${jobId}/eligibility`, {
    headers: authHeaders(),
    cache: "no-store",
  });

  const data = await response.json().catch(() => null);

  if (!response.ok) {
    const message =
      typeof data?.detail === "string"
        ? data.detail
        : "Unable to check eligibility.";

    throw new Error(message);
  }

  return data as JobEligibilityResult;
}

export type AtsAlignmentStatus = "matched" | "partial" | "missing";

export type AtsRequirementCategory = "must_have" | "preferred";

export type AtsRequirementResult = {
  requirement_id: string;
  requirement_type: "skill" | "experience" | "education" | "certification";
  category: AtsRequirementCategory;
  requirement_text: string;
  status: AtsAlignmentStatus;
  jd_evidence: string;
  resume_evidence: string | null;
  explanation: string;
  confidence: "high" | "medium" | "low";
};

export type AtsScoreComponent = {
  name: string;
  weight: number;
  score: number;
  weighted_score: number;
  explanation: string;
};

export type AtsAlignmentResult = {
  id: string;
  job_id: string;
  resume_version_id: string;
  job_intelligence_id: string;
  engine_version: string;
  overall_score: number;
  confidence: "high" | "medium" | "low";
  scoring_version: string;
  must_have_total: number;
  must_have_matched: number;
  preferred_total: number;
  preferred_matched: number;
  must_have_ceiling: number | null;
  score_components: AtsScoreComponent[];
  requirement_results: AtsRequirementResult[];
  created_at: string;
};

// resumeVersionId is optional: when omitted, the query param is left off
// entirely and the backend falls back to its existing default resume
// resolution (most recent Resume, preferring its master ResumeVersion) -
// AJI-019 never changes that default, it only lets the caller override it.
export async function calculateAtsAlignment(
  jobId: string,
  resumeVersionId?: string,
): Promise<AtsAlignmentResult> {
  const searchParams = new URLSearchParams();
  if (resumeVersionId) {
    searchParams.set("resume_version_id", resumeVersionId);
  }
  const query = searchParams.toString();

  const response = await fetch(
    `${API_BASE_URL}/jobs/${jobId}/ats${query ? `?${query}` : ""}`,
    {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        ...authHeaders(),
      },
      cache: "no-store",
    },
  );

  const data = await response.json().catch(() => null);

  if (!response.ok) {
    const message =
      typeof data?.detail === "string"
        ? data.detail
        : "Unable to calculate ATS Alignment.";

    throw new Error(message);
  }

  return data as AtsAlignmentResult;
}

export async function calculateJobMatch(
  jobId: string,
  resumeVersionId?: string,
): Promise<JobMatchResult> {
  const token =
    typeof window !== "undefined"
      ? localStorage.getItem("ai_job_intelligence_token")
      : null;

  const searchParams = new URLSearchParams();
  if (resumeVersionId) {
    searchParams.set("resume_version_id", resumeVersionId);
  }
  const query = searchParams.toString();

  const response = await fetch(
    `${API_BASE_URL}/jobs/${jobId}/match${query ? `?${query}` : ""}`,
    {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        ...(token
          ? {
              Authorization: `Bearer ${token}`,
            }
          : {}),
      },
      cache: "no-store",
    },
  );

  const data = await response.json().catch(() => null);

  if (!response.ok) {
    const message =
      typeof data?.detail === "string"
        ? data.detail
        : "Unable to calculate job match.";

    throw new Error(message);
  }

  return data as JobMatchResult;
}

export type GapSuggestionType =
  | "ADD_IF_TRUE"
  | "REPHRASE_EXISTING"
  | "HIGHLIGHT_EXISTING";

export type GapSuggestion = {
  requirement_id: string;
  requirement_type: "skill" | "experience" | "education" | "certification";
  category: AtsRequirementCategory;
  requirement_text: string;
  status: "partial" | "missing";
  jd_evidence: string;
  resume_evidence: string | null;
  suggestion_type: GapSuggestionType;
  explanation: string;
  explanation_source: "ai" | "deterministic";
  suggestion_text: string;
  suggestion_source: "ai" | "deterministic";
  confidence: "high" | "medium" | "low";
};

export type GapAnalysisResult = {
  id: string;
  job_id: string;
  resume_version_id: string;
  job_intelligence_id: string;
  ats_alignment_id: string;
  analysis_version: string;
  analyzer_version: string;
  prompt_version: string;
  model_provider: string | null;
  model_name: string | null;
  generation_status: string;
  must_have_gap_count: number;
  preferred_gap_count: number;
  gaps: GapSuggestion[];
  created_at: string;
};

export async function getGapAnalysis(
  jobId: string,
  resumeVersionId?: string,
): Promise<GapAnalysisResult> {
  const searchParams = new URLSearchParams();
  if (resumeVersionId) {
    searchParams.set("resume_version_id", resumeVersionId);
  }
  const query = searchParams.toString();

  const response = await fetch(
    `${API_BASE_URL}/jobs/${jobId}/gap-analysis${query ? `?${query}` : ""}`,
    {
      headers: authHeaders(),
      cache: "no-store",
    },
  );

  const data = await response.json().catch(() => null);

  if (!response.ok) {
    const message =
      typeof data?.detail === "string"
        ? data.detail
        : "Unable to load Gap Analysis.";

    throw new Error(message);
  }

  return data as GapAnalysisResult;
}

export async function calculateGapAnalysis(
  jobId: string,
  resumeVersionId?: string,
): Promise<GapAnalysisResult> {
  const searchParams = new URLSearchParams();
  if (resumeVersionId) {
    searchParams.set("resume_version_id", resumeVersionId);
  }
  const query = searchParams.toString();

  const response = await fetch(
    `${API_BASE_URL}/jobs/${jobId}/gap-analysis${query ? `?${query}` : ""}`,
    {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        ...authHeaders(),
      },
      cache: "no-store",
    },
  );

  const data = await response.json().catch(() => null);

  if (!response.ok) {
    const message =
      typeof data?.detail === "string"
        ? data.detail
        : "Unable to calculate Gap Analysis.";

    throw new Error(message);
  }

  return data as GapAnalysisResult;
}

// ---------------------------------------------------------------------------
// Resume Improvement Approval & Recheck (AJI-021)
// ---------------------------------------------------------------------------

export type ImprovementDecisionAction = "approve" | "skip";

/**
 * One submitted decision. There is deliberately no `suggestion_type` or
 * `applied_text` field: the backend reads the suggestion type from the
 * stored Gap Analysis row and writes only `user_content`, and rejects a
 * request carrying either field outright.
 */
export type ImprovementDecisionInput = {
  requirement_id: string;
  action: ImprovementDecisionAction;
  truth_confirmed?: boolean;
  user_content?: string | null;
};

export type ImprovementDecisionRecord = {
  requirement_id: string;
  requirement_text: string;
  category: AtsRequirementCategory;
  suggestion_type: GapSuggestionType;
  action: ImprovementDecisionAction;
  truth_confirmed: boolean;
  applied_text: string | null;
  content_source: "user" | "none";
};

export type RequirementTransitionDirection =
  | "improved"
  | "unchanged"
  | "regressed"
  | "added"
  | "removed";

export type RequirementTransition = {
  requirement_id: string;
  requirement_text: string;
  category: AtsRequirementCategory;
  before_status: AtsAlignmentStatus | null;
  after_status: AtsAlignmentStatus | null;
  direction: RequirementTransitionDirection;
  was_approved: boolean;
};

export type ImprovementComparison = {
  baseline_ats_alignment_id: string;
  baseline_resume_version_id: string;
  baseline_score: number;
  baseline_must_have_matched: number;
  baseline_must_have_total: number;
  baseline_preferred_matched: number;
  baseline_preferred_total: number;
  recheck_ats_alignment_id: string;
  recheck_resume_version_id: string;
  recheck_score: number;
  recheck_must_have_matched: number;
  recheck_must_have_total: number;
  recheck_preferred_matched: number;
  recheck_preferred_total: number;
  score_delta: number;
  must_have_delta: number;
  preferred_delta: number;
  improved_count: number;
  unchanged_count: number;
  regressed_count: number;
  transitions: RequirementTransition[];
};

export type ResumeImprovementResult = {
  id: string;
  job_id: string;
  gap_analysis_id: string;
  baseline_ats_alignment_id: string;
  parent_resume_version_id: string;
  child_resume_version_id: string;
  child_resume_version_name: string;
  engine_version: string;
  approved_count: number;
  skipped_count: number;
  recheck_status: "pending" | "complete" | "failed";
  recheck_error: string | null;
  recheck_ats_alignment_id: string | null;
  decisions: ImprovementDecisionRecord[];
  // null whenever the recheck has not produced a result - the created
  // version is still reported in full, because a failed recheck never
  // costs the user the version they approved.
  comparison: ImprovementComparison | null;
  created_at: string;
  updated_at: string;
};

/**
 * The backend returns `{ code, message }` for this feature's own errors
 * so the UI can distinguish "you still need to confirm this is true"
 * from a transport failure without matching on prose.
 */
export class ResumeImprovementError extends Error {
  code: string;

  constructor(message: string, code: string) {
    super(message);
    this.name = "ResumeImprovementError";
    this.code = code;
  }
}

function toImprovementError(data: unknown, fallback: string): Error {
  const detail = (data as { detail?: unknown } | null)?.detail;

  if (detail && typeof detail === "object" && "message" in detail) {
    const { message, code } = detail as { message?: string; code?: string };
    return new ResumeImprovementError(message || fallback, code || "error");
  }

  if (typeof detail === "string") {
    return new ResumeImprovementError(detail, "error");
  }

  return new ResumeImprovementError(fallback, "error");
}

export async function getResumeImprovement(
  jobId: string,
  resumeVersionId?: string,
): Promise<ResumeImprovementResult> {
  const searchParams = new URLSearchParams();
  if (resumeVersionId) {
    searchParams.set("resume_version_id", resumeVersionId);
  }
  const query = searchParams.toString();

  const response = await fetch(
    `${API_BASE_URL}/jobs/${jobId}/resume-improvement${query ? `?${query}` : ""}`,
    {
      headers: authHeaders(),
      cache: "no-store",
    },
  );

  const data = await response.json().catch(() => null);

  if (!response.ok) {
    throw toImprovementError(data, "Unable to load your approved improvements.");
  }

  return data as ResumeImprovementResult;
}

export async function createResumeImprovement(
  jobId: string,
  gapAnalysisId: string,
  decisions: ImprovementDecisionInput[],
): Promise<ResumeImprovementResult> {
  const response = await fetch(
    `${API_BASE_URL}/jobs/${jobId}/resume-improvement`,
    {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        ...authHeaders(),
      },
      body: JSON.stringify({
        gap_analysis_id: gapAnalysisId,
        decisions,
      }),
      cache: "no-store",
    },
  );

  const data = await response.json().catch(() => null);

  if (!response.ok) {
    throw toImprovementError(
      data,
      "Unable to apply your approved improvements.",
    );
  }

  return data as ResumeImprovementResult;
}

export async function retryResumeImprovementRecheck(
  jobId: string,
  improvementId: string,
): Promise<ResumeImprovementResult> {
  const response = await fetch(
    `${API_BASE_URL}/jobs/${jobId}/resume-improvement/${improvementId}/recheck`,
    {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        ...authHeaders(),
      },
      cache: "no-store",
    },
  );

  const data = await response.json().catch(() => null);

  if (!response.ok) {
    throw toImprovementError(data, "Unable to recheck your new version.");
  }

  return data as ResumeImprovementResult;
}
